Add PackNuspec MSBuild task

natemcmaster opened the pull request:
Adds an MSBuild task that can pack a .nuspec file.

Adapted from https://github.com/aspnet/BuildTools/blob/dev/modules/KoreBuild.Tasks/PackNuSpec.cs

natemcmaster commented:
So, to close the loop on this, is this task, in the form I have proposed, something we want to move forward with? Or should are there enough concerns about its API shape, functionality, and relationship to NuGet's official PackTask that we should table for now?

natemcmaster commented:
In other words, is this something dotnet/corefx, dotnet/core-setup, and others could adopt? The motivation here was to move this task about of aspnet and into arcade.

jaredpar commented:
I'd prefer to proceed. It doesn't seem like NuGet is willing to set a date for allowing `dotnet pack nuspec` file. The alternative approach they suggest is burdensome. I'd take this type of solution until NuGet can get back to the needed feature set.

weshaggard commented:
I'm fine with proceeding as well. We just need to be able to react and move to officially shipping versions of tools if they come online to provide what we need. That goes not just for this task but all the tools we are putting in arcade. In an ideal world arcade would be empty and we use only shipping tools.

rrelyea commented:
@weshaggard / @natemcmaster / @jaredpar - we'd like people to start moving their main projects to include pack metadata, rather than having it external from msbuild.
While we can add a feature to "dotnet pack" to provide support for packing a nuspec directly, we are not convinced that investment is a great use of our time.
A) why can you not make your existing project file (and props/targets) provide the correct metadata for pack?
@rohit21agrawal

jaredpar commented:
> While we can add a feature to "dotnet pack" to provide support for packing a nuspec directly, we are not convinced that investment is a great use of our time.

There are NuPkg types which simply cannot be expressed as a project file. Take the MSBuild NuPkg the Roslyn compiler emits, and virtually any other MSBuild + Tool based package. In that case we have to create project, which will never build and will never ship , yet must still restore, simply to use a NuSpec file as we've been doing for the past five+ years. 

The other motivation is conversion pain. Lots of code out there that uses `nuget pack <nuspec>` today that needs to change pretty significantly in order to use this new model. It's yet another pain point in adopting the CLI.

natemcmaster commented:
Updated this PR. Rebased on master and renamed the api to be more consistent about naming of file paths and directories.

rohit21agrawal commented:
@jaredpar you can customize your package as much as you like using a SDK or classic csproj. I realize there is a documentation gap , so I have filed another PR : https://github.com/NuGet/docs.microsoft.com-nuget/pull/794

rohit21agrawal commented:
@natemcmaster wouldn't an easier approach for you be create a new dotnet project behind the scenes and pack a nuspec using that? the end user doesn't have to know what happened. it will be lesser code for you to maintain.

natemcmaster commented:
> wouldn't an easier approach for you be create a new dotnet project behind the scenes and pack a nuspec using that

No, it is definitely not easier. I feel like I've said this a dozen times now and its not being heard. I want to pack with just a .nuspec not a .csproj file! Here are some of the reasons why I gave up on the csproj -> nuspec generation. Some issues were closed as "fixed", but I still think the MSBuild property names and item groups you have to use in order to get the package layout you want are poorly named and hard to work with. .nuspec is so much simpler.

https://github.com/NuGet/Home/issues/5063
https://github.com/NuGet/Home/issues/4254
https://github.com/NuGet/Home/issues/6653
https://github.com/NuGet/Home/issues/6351
https://github.com/NuGet/Home/issues/4704

rohit21agrawal commented:
>  I feel like I've said this a dozen times now and its not being heard. I want to pack with just a .nuspec not a .csproj file!

You are again misunderstanding me. I am not saying use a csproj file everytime you have to pack a nuspec - I am saying write a task that does that for you automatically. If I wrote a top level command ``` dotnet packnuspec foo.nuspec``` that created a csproj file behind the scenes, and ran ``` dotnet pack temp.csproj /p:<your nuspec file> ``` behind the scenes, would you care that it was implemented that way? The end result is that the nuspec got packed.

natemcmaster commented:
Thanks for clarifying. I misunderstood that's what you meant by "create a new dotnet project behind the scenes". 

I'm not convinced this wouldn't be any easier than using NuGet's API directly. There is a lot to handle in the approach of shelling out to `dotnet pack`.  Argument escaping, detecting and using the right `dotnet.exe` or `msbuild.exe` executable, flowing errors back to the MSBuild logger, generating and cleaning up temp files, etc. You can't execute /t:Pack without first restoring `temp.csproj`, you can't use [empty replacement tokens](https://github.com/NuGet/Home/issues/6722), you would end up using the NuGet API anyways to detect cases [like this](https://github.com/dotnet/arcade/blob/namc/nuget-tasks/src/Microsoft.DotNet.Build.Tasks.NuGet/src/PackNuspec.cs#L131-L134), or to generate and group [dependencies](https://github.com/dotnet/arcade/blob/namc/nuget-tasks/src/Microsoft.DotNet.Build.Tasks.NuGet/src/PackNuspec.cs#L247-L280) or [package files](https://github.com/dotnet/arcade/blob/namc/nuget-tasks/src/Microsoft.DotNet.Build.Tasks.NuGet/src/PackNuspec.cs#L225-L245) from MSBuild items. I'm not sure we gain much. 

That said, is there some obvious trouble we're going to hit in the future? Are you planning to remove PackageBuilder APIs or make changes that would break this task? The PackNuspec task in this PR is basically the same code we've been using in aspnet for quite a while and it's worked great for us. I wanted to move it to arcade so others can use it too.

natemcmaster commented:
@rrelyea @nkolev92 is this code something you would accept as a contribution to NuGet itself? My concern here is that (1) MSBuild does not isolate task dependencies and (2) the NuGet public API regularly takes breaking changes https://github.com/NuGet/Home/issues/6398. 

If not, I think it's worth waiting on this until we address https://github.com/dotnet/arcade/issues/47 before we take this in.

rohit21agrawal commented:
@natemcmaster  given that nuget.exe is not cross-platform, i think what you want is to have a packnuspec command in dotnet.exe . @livarcocc will be the right person to decide if they want to take it. However, I will let @rrelyea and @livarcocc decide who is going to own it if they do decide to take it in dotnet cli.

jaredpar commented:
@rohit21agrawal 

>  If I wrote a top level command dotnet packnuspec foo.nuspec that created a csproj file behind the scenes, and ran dotnet pack temp.csproj /p:<your nuspec file> behind the scenes, would you care that it was implemented that wa

I would be concerned given the way that it's currently implemented. Today it is necessary to run `dotnet restore` in order to pack the project in this manner. This will have an observable output on our build in terms of what artifacts are written to disk (adds data to the intermediate directory) and it introduces an extra network round trip in our process that is strictly not needed. 

The extra artifact complaint is not being pedantic, it's about learning from past problems with other tools: 

1. The project you create must have a unique name in order to avoid restore conflicts with other projects in our solution. In order to do that you will essentially name the project `Guid.NewGuid().ToString("N")`. That is problematic also because it means the intermediate directory grows with every build as the project files are created.
1. The use of a unique project name also creates determinism issues. This is a priority for the division and company. Any time generated names enter a build it hampers determinism. 
1. The project file needs to obey the rules of our build system: picking up the correct location for `bin` and `obj` directories for example. Given the many different ways builds can hook these values there really isn't a way to generalize this. You have to have specific knowledge about the build in order to hook your project up to it correctly. 

Generating a project on the fly is not an implementation detail. It's an observable artifact of build that we'd likely have to reject. 

> The end result is that the nuspec got packed.

The end result is just as important as the work needed to get there.

markwilkie commented:
cc/ @tmat for visibility.

@natemcmaster - any recommend for next steps on this?

natemcmaster commented:
I recommend we pursue getting this merged. There are still some implementation challenges, mostly due to the nature of assembly loading in MSBuild, but I think getting this API available is a good first step.

Not sure who else should review the code. I think I addressed most of the feedback given so far.

markwilkie commented:
I added @tmat to the reviewers list.   @weshaggard, @jaredpar, @chcosta  - could we can an ACK or things which should be addressed prior to merging?

tmat commented on src/Tasks.Shared/ArcadeTaskErrors.cs:
```diff
@@ -0,0 +1,12 @@
+{
+    internal static class ArcadeTaskErrors
+    {
+        public const string InvalidNuspecFile = "ARCADE1001";
```
>ARCADE [](start = 49, length = 6)

These errors seem to be nuspec-specific, why not use NUSPEC prefix?

natemcmaster replied:
Good feedback. I'll update.

tmat commented on src/Microsoft.DotNet.Build.Tasks.NuGet/targets/sdk/Sdk.props:
```diff
@@ -0,0 +1,3 @@
+<Project ToolsVersion="14.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003" />
+    <Import Project="..\build\Microsoft.DotNet.Build.Tasks.NuGet.props" />
+</Project>
```
I don't think this package needs to be an SDK.

natemcmaster replied:
Why not be an SDK package? The goal is to simplify acquisition and usage for as many consumers as possible.

tmat replied:
What is simpler about SDK package?

natemcmaster replied:
SDK packages make it easier to ensure a unified version of the package is shared throughout the MSBuild process. The version, specified in global.json, is applied to all projects. When a task is pulled as a PackageReference there is opportunity for two projects in the same solution to use a different version. We've run into issues in the past when using PackageReference where MSBuild tries (and fails) to load the same task from two different assemblies.

tmat commented on src/Microsoft.DotNet.Build.Tasks.NuGet/Microsoft.DotNet.Build.Tasks.NuGet.csproj:
```diff
@@ -0,0 +1,32 @@
+    <Reference Include="Microsoft.Build.Tasks.v4.0" />
+    <Reference Include="Microsoft.Build.Framework" />
+    <Reference Include="Microsoft.Build.Utilities.v4.0" />
+  </ItemGroup>
```
Why not just use 
<PackageReference Include="Microsoft.Build" Version="$(MicrosoftBuildVersion)" />
    <PackageReference Include="Microsoft.Build.Tasks.Core" Version="$(MicrosoftBuildTasksCore)" />

with 

    <MicrosoftBuildVersion>15.6.82</MicrosoftBuildVersion> 
    <MicrosoftBuildTasksCore>15.6.82</MicrosoftBuildTasksCore>

natemcmaster replied:
MSBuild 14 support.

tmat replied:
Why do we need msbuild 14 support? We should be using 15.6. RepoToolset requires 15.6

natemcmaster replied:
My understanding per [this spec](https://github.com/dotnet/arcade/blob/master/Documentation/TaskPackages.md) was that our intention was to create simple tasks available in packages that are generally useful to the .NET team. My intention here is to provide an MSBuild task available to many consumers. RepoToolset is one of those consumers, but not the be-all of Arcade.

And yes, aspnet has some MSBuild 14 stuff, sadly.

tmat replied:
That's true, however I'd rather ask everyone to use the latest versions of our tools if possible (including MSBuild) and kept our infrastructure simple. 

What is the reason ASP.NET needs MSBuild 14 for building? Is there an issue to remove that requirement?

tmat commented:
Related PR in RepoToolset: https://github.com/dotnet/roslyn-tools/pull/235
Once this task is available from Arcade we can hook it up in Roslyn Tools as well.

tmat commented:
This might have been discussed previously, but I couldn't find answer -- why not implement an alternative to `GenerateNuspec` task (`GenerateNuspec2`) that would override the one in nuget targets if the projects imports the Arcade package and leave creating the actual package to NuGet Pack? It would be easier to converge that eventually with the `GenerateNuspec` task. Is there something that wouldn't work using this approach?

natemcmaster commented on src/Microsoft.DotNet.Build.Tasks.NuGet/src/PackNuspec.cs:
```diff
@@ -0,0 +1,282 @@
+        /// <summary>
+        /// Subsitution in the nuspec via $key$.
+        /// </summary>
+        public string[] Properties { get; set; }
```
Not opposed to this, but can we add this as a separate PR? This one has gone on long enough, and I think what you're proposing here can be done after the task itself has been added.

tmat replied:
Sure.

natemcmaster commented:
> why not implement an alternative to GenerateNuspec task (GenerateNuspec2)

The goal of this PR is simply to add a task for packing a nuspec, not targets to replace `/t:pack` The intention is to provide support for creating packages when we don't want to or can't generate the nuspec from a .csproj

natemcmaster commented:
Updated with feedback.

tmat commented:
> The intention is to provide support for creating packages when we don't want to or can't generate the nuspec from a .csproj

I get that but why not add a task instead that allows us to generate nuspec from a project, filling in the pieces that are currently missing, that are the reason why you can't use the default targets?

tmat commented:
Are yous saying you don't even want to have to create a project file per nuget package?

natemcmaster commented:
> Are you saying you don't even want to have to create a project file per nuget package?

Yes.

tmat commented:
I see. Do you see that changing in future and having this task only temporarily until we update the repos to have a project per nuget?

natemcmaster commented:
I see it as a permanent, unless we intend to adopt [nuproj](https://github.com/nuproj/nuproj) as a semi-official .NET thing (which would be cool). 

Primary use case today in aspnet are packages like libuv or Microsoft.AspNetCore.Server.IIS that only contain native assemblies, or internal packages that contain metadata for automation purposes only.

tmat commented:
> I see it as a permanent, unless we intend to adopt nuproj as a semi-official .NET thing (which would be cool).

@natemcmaster We don't need nuproj (although perhaps we should explore it). A regular C# project with [a couple of build targets](https://github.com/dotnet/roslyn-analyzers/blob/master/nuget/Directory.Build.targets#L10-L14) suppressed work pretty well. For example: https://github.com/dotnet/roslyn-analyzers/tree/master/nuget

natemcmaster commented:
Closing this for now. As discussed in Arcade sync up, we still think there is a dire need for a PackNuspec task in MSBuild, however, it appears aspnet, corefx, roslyn, and others have already worked around this with their own versions of a PackNuspec task. We want to unify those one task at some point, but would like time to better design something that makes the various consumers happy. Also, we'd love to work with the NuGet team to make this happen. I have opened https://github.com/dotnet/arcade/issues/126 to follow up on this.
